### DISCUSSION TITLE
Possible over-optimization

### DISCUSSION THREAD

**Discussion by jeremyckahn** [jeremyckahn/shifty#128] [General]
I uploaded [this video](https://www.youtube.com/watch?v=dyPc0qedcrI&t=737) detailing the performance improvements to Shifty in 2.14.0. At the linked point in the video, I explain the reused variables used in Shifty's private helpers. For instance, here: https://github.com/jeremyckahn/shifty/blob/af71d52777491493b89bd5aa35d63a8ffc015707/src/tweenable.js#L54-L98

This is an admittedly weird way to write JavaScript, but the basic idea is that these functions' variables are being reused from invocation to invocation. I assumed this would be faster, as the variables are being reused rather than redeclared many times in a tick. However, it's not so cut and dry. Take a look at [this Gist that benchmarks this technique](https://gist.github.com/jeremyckahn/b97f326f7534cca4585322db9f6db040#file-benchmark-js), and its results:

```
Regular function:  Completed 10 iterations in 0ms
Closured function: Completed 10 iterations in 0ms

Regular function:  Completed 100 iterations in 0ms
Closured function: Completed 100 iterations in 0ms

Regular function:  Completed 10000 iterations in 3ms
Closured function: Completed 10000 iterations in 2ms

Regular function:  Completed 1000000 iterations in 27ms
Closured function: Completed 1000000 iterations in 34ms

Regular function:  Completed 100000000 iterations in 1758ms
Closured function: Completed 100000000 iterations in 3403ms
```

The "Closured" function (the one with the reused variables) is slightly faster for fewer iterations, but starts to become slower for greater numbers of them. This is the **opposite** of what I expected.

I'll experiment a bit with which approach yields better results in Shifty. If it seems that the standard form is faster, I'll switch to that as the code would be far more maintainable. I'll update this thread with what I learn!

**Comment by jeremyckahn** [comment]
Behold, science!

I made an experimental branch to explore this idea a bit. Here's the relevant commit: https://github.com/jeremyckahn/shifty/commit/db7c646219e5fbaa7ac9dcf53e36238cf9f49658?w=1

And here are benchmarks for the before each approach:

  - [2.14.1 with reused variables](https://cdpn.io/jeremyckahn/debug/QWKpdYv/bZMQWKwRwKxA)
  - [2.14.1 experimental build with redeclared variables](https://cdpn.io/jeremyckahn/debug/NWRpded/vWkRwGNmNoKM)

And here's how the memory utilization compares (experimental build is on the right):

<img width="2559" alt="Screen Shot 2020-12-14 at 9 45 55 PM" src="https://user-images.githubusercontent.com/366330/102168952-c9e0e200-3e56-11eb-94b6-99f3eeae027f.png">

The version with re-declared variables is **trivially more efficient**, both with with CPU and memory usage. The code is also far more readable and maintainable. Indeed, I got too clever and over-optimized here. But now we have the metrics to prove it!

I'll release the improved version as 2.14.2 shortly. 🙂

**Comment by jeremyckahn** [comment]
[2.14.2](https://github.com/jeremyckahn/shifty/releases/tag/v2.14.2) has been shipped!

**Comment by 4ian** [comment]
Hey @jeremyckahn I hope it's ok to revive this discussion now - just to tell you thank you for posting this and your observations/benchmarks! Very useful to understand this kind of over optimizations and how they scale/don't scale.
Seems that JS engines are working well with "traditional" variables declared on the stack, that's good to know! :) 

Thanks again for your work on Shifty and on these kind of optimisations!

**Comment by jeremyckahn** [comment]
I'm glad to hear you got something out of this exploration as well, @4ian! It was fun to go super deep on performance analysis like this. I don't often get to do that! 😄
